File: Hash_Table/problem30/problem30.cpp
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        if (words.empty() || s.empty()) {
            return {};
        }
        const int wlen = static_cast<int>(words[0].size());
        const int n = static_cast<int>(words.size());
        const int total = wlen * n;
        if (static_cast<int>(s.size()) < total) {
            return {};
        }

        unordered_map<string, int> need;
        need.reserve(words.size() * 2);
        for (const string& w : words) {
            ++need[w];
        }

        vector<int> ans;

        for (int offset = 0; offset < wlen; ++offset) {
            unordered_map<string, int> cur;
            int left = offset;
            int matched = 0;
            for (int j = offset; j + wlen <= static_cast<int>(s.size()); j += wlen) {
                const string word = s.substr(j, wlen);
                auto it = need.find(word);
                if (it != need.end()) {
                    ++cur[word];
                    ++matched;
                    while (cur[word] > it->second) {
                        const string lw = s.substr(left, wlen);
                        if (--cur[lw] == 0) {
                            cur.erase(lw);
                        }
                        --matched;
                        left += wlen;
                    }
                    if (matched == n) {
                        ans.push_back(left);
                        const string lw = s.substr(left, wlen);
                        if (--cur[lw] == 0) {
                            cur.erase(lw);
                        }
                        --matched;
                        left += wlen;
                    }
                } else {
                    cur.clear();
                    matched = 0;
                    left = j + wlen;
                }
            }
        }

        return ans;
    }
};

File: Hash_Table/problem30/problem30.py
from collections import Counter
from typing import List


class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not words or not s:
            return []
        wlen = len(words[0])
        n = len(words)
        total = wlen * n
        if len(s) < total:
            return []

        need: Counter[str] = Counter(words)
        out: list[int] = []

        for offset in range(wlen):
            left = offset
            cur: Counter[str] = Counter()
            matched = 0
            for j in range(offset, len(s) - wlen + 1, wlen):
                word = s[j : j + wlen]
                if word in need:
                    cur[word] += 1
                    matched += 1
                    while cur[word] > need[word]:
                        lw = s[left : left + wlen]
                        cur[lw] -= 1
                        if cur[lw] == 0:
                            del cur[lw]
                        matched -= 1
                        left += wlen
                    if matched == n:
                        out.append(left)
                        lw = s[left : left + wlen]
                        cur[lw] -= 1
                        if cur[lw] == 0:
                            del cur[lw]
                        matched -= 1
                        left += wlen
                else:
                    cur.clear()
                    matched = 0
                    left = j + wlen

        return out
